Hunt System

Apocarypse opened the pull request:
<!-- place 'x' mark between square [] brackets to affirm: -->
**_I affirm:_**
- [x] that I agree to Project Topaz's [Limited Contributor License Agreement](http://project-topaz.com/blob/release/CONTRIBUTOR_AGREEMENT.md), as written on this date
- [x] that I've _tested my code_ since the last commit in the PR, and will test after any later commits

Functional hunt system implementation. The only thing missing is Evoliths, which require synergy to be working. The VNMs also do not have scripts, as they are unimplemented at this time.

Please do not squash when merging this PR as there are collaborative commits which should remain separate.

I would like to thank Kreidos for all his help on this project. He was there to assist every step of the way and definitely deserves half the credit.

Fixes #240

Kreidos committed:
Hunt System NM Scripts

Apocarypse committed:
hunt system implementation

cocosolos commented on scripts/globals/hunts.lua:
```diff
@@ -0,0 +1,1616 @@
+  local registryZone = zone[player:getZoneID()]
+
+  -- one vana'diel day lockout timer after completing a hunt
+  if (VanadielYear() * 360) + VanadielDayOfTheYear() < player:getCharVar("[hunt]nextHunt") then
```
Could use `if os.time() < player:getCharVar("[hunt]nextHunt") then`

Apocarypse replied:
how about `if player:getCharVar("[hunt]nextHunt") < getVanaMidnight()then`

Kreidos replied:
I like this idea. Ideally we should compare against a consistent source.

cocosolos commented on scripts/globals/hunts.lua:
```diff
@@ -0,0 +1,1616 @@
+  -- cancels training regime and clears all vars
+  elseif option == 4 then
+    player:messageSpecial(msg[4])
+    tpz.regime.clearPlayerVars(player)
```
Was this changed to `clearRegimeVars`?

Apocarypse replied:
Yes! I'm glad you caught that.

cocosolos commented on scripts/globals/hunts.lua:
```diff
@@ -0,0 +1,1616 @@
+    player:messageSpecial(msg[6], scyldBounty, scylds)
+    player:addCurrency("scyld", scyldBounty)
+    -- give player evoliths here
+    player:setCharVar("[hunt]nextHunt", (VanadielYear() * 360) + (VanadielDayOfTheYear() + 1))
```
Could use `player:setCharVar("[hunt]nextHunt", getVanaMidnight())`

Apocarypse committed:
changed to 4 spaces, fixed currency messages, and properly renamed vars

cocosolos commented on scripts/globals/hunts.lua:
```diff
@@ -0,0 +1,1616 @@
+  local registryZone = zone[player:getZoneID()]
+
+  -- one vana'diel day lockout timer after completing a hunt
+  if player:getCharVar("[hunt]nextHunt") < getVanaMidnight() then
```
This should be checking `os.time()`, reason being that `getVanaMidnight()` very slightly varies in what it returns, even for the same day, so if you're lucky then `getVanaMidnight` will return a midnight that's maybe 1s before the `getVanaMidnight` that is saved in the char var. This should check `if os.time() < player:getCharVar("[hunt]nextHunt") then` (it's also checking the requirement backwards, this block should execute if the char is blocked)

Kreidos replied:
I have a question. Since os.time() is lua native it always returns the system time; whereas getVanaMidnight ties into the time calculations and so could be affected by any offsets which are (or might) be implemented in the core. Couldn't that pose an issue to using os.time() which is completely independent of server time calculations?

cocosolos replied:
It's still using getVanaMidnight to set the time, so that's all accounted for. This check is just seeing if the current time is past what has been determined is the next vana midnight when that var was set.

cocosolos commented on scripts/globals/hunts.lua:
```diff
@@ -0,0 +1,1616 @@
+        [1577] = { pageId = 197, mobId = 17211561, bounty =  5, fee =   0 }, -- Stinging Sophie
+        [1585] = { pageId = 198, mobId = 17211714, bounty =  5, fee =   0 }, -- Maighdean Uaine
+        [1593] = { pageId = 199, mobId = 17211666, bounty =  5, fee =   0 }, -- Bedrock Barry
+        [1601] = { pageId = 200, mobId = 17215888, bounty =  5, fee =   0 }, -- Leaping Lizzy
```
I'm not sure if this is very common, but I happened to test this on Lizzy and she actually has 2 IDs, so might need to add the other one `17215868`.

Apocarypse replied:
I asked about this and got no reply unfortunately. There are some NMs that have a single spawn point under one ID, and like 50 spawn points under another ID. In those cases, I opted for the 50. I don't know what the logic is there exactly.

TeoTwawki replied:
~~tabling all the IDs of all these NMs is the wrong approach. Followup what all existing regimes do and include the regime ID at the mob, then match on regime ID instead of performing a lookup to see if that specific mob is in a table. We don't need to keep track of them all because its never going to change, and you already have that line there anyway. model it after the existing regimes instead of trying to centralize every bit of data to one place as this is a case where that just gets in the way.~~

~~`tpz.hunts.checkHunt(mob, player, isKiller)`
compare with
`tpz.regime.checkRegime(player, mob, 80, 1, tpz.regime.type.FIELDS)`
and
`tpz.regime.checkRegime(player, mob, 640, 2, tpz.regime.type.GROUNDS)`
you're trying to reinvent the wheel to separate hunts from these (understandable, a lot of code doesn't apply here) but missed why the regime ID was included.~~

now looking at this:
`tpz.hunts.checkHunt(mob, player, isKiller)`
We should not need isKiller at all as in retail all member who have the hunt regime active get credit and its not going to stack up (onMobDeath code will run once-per-member), so I suggest just go with:
`tpz.hunts.checkHunt(mob, player, RegimeIDhere)` to minimize your changes.

Edit: this could have been worded better by me, see coco's post which got straight to what it'd look like and didn't risk misunderstandings

TeoTwawki replied:
p.s. if you're thinking "but then we need to keep track of which NMs go in which regime!" no, because the regime IDs will never shift. set once done for life. :)

Kreidos replied:
I gotta disagree here on part of that. isKiller isn't necessary, you're right, but frankly I think having all the mobs tabled here with a simple lookup to a table is a much faster and better method than breaking out every ID into separate mob scripts. Regimes takes the ID it gets and combs through a table using a for loop anyway to find the mob which is far from elegant. This solution involves one direct check to one table and I gotta go with the performance method here.

Kreidos replied:
I think this works well, and if you believe you have a better solution I welcome you to PR the files after this merges. :)

TeoTwawki replied:
yeah no, I'm no opening a competing PR just because you prefer listing every NMs ID into the table and then trying to make workaround for the inherent shortcomings of that approach. 

A given NM can have multiple IDs. at a minimum would would need multiple ID entries. There is no reason we could not match on regime ID instead of mobID, loop or no loop.

cocosolos replied:
Switching to `tpz.hunts.checkHunt(mob, player, RegimeIDhere)` would actually solve the multiple ID mobs issue all at once and would require minimal changes. `checkHunt` becomes:
```lua
    function tpz.hunts.checkHunt(mob, player, RegimeIDhere)
        local huntZone = player:getCharVar("[hunt]zone")
        local huntId = player:getCharVar("[hunt]id")
        local playerHunt = zone[huntZone].hunt[huntId]

        -- required NM has been defeated
        if player:getCharVar("[hunt]status") == 1 and huntId == RegimeIDhere then
           player:messageBasic(tpz.msg.basic.FOV_DEFEATED_TARGET + 20)
           player:setCharVar("[hunt]status", 2)
        end
    end
```
but that would also require removing all the mob IDs from the tables and updating all the mob files as well.

TeoTwawki replied:
@cocosolos that is what I was going for. It does suck to re-edit all those mob scripts, but they all did have to have an edit to start with so its not like these didn't have code going to them. That said I will volunteer to make the mob script changes if apoc wants to take this road.

the alternative is harder - alterations to support multiple mobID elements and then the inevitable future problems that come with mobIDs later


((if anyone can come up with a 3rd option, I'm all ears)

cocosolos commented on scripts/globals/hunts.lua:
```diff
@@ -0,0 +1,1616 @@
+    tpz.hunts.clearHuntVars(player)
+
+    -- scylds cap at 1000
+    if player:getCurrency("scyld") + bounty > 1000 then
```
`bounty` is nil, is this supposed to be `scyldBounty`?

Apocarypse replied:
Yes, I think you're right. I'll change it.

cocosolos commented on scripts/globals/hunts.lua:
```diff
@@ -0,0 +1,1616 @@
+
+    player:messageSpecial(msg[5])
+    player:messageSpecial(msg[6], scyldBounty, scylds)
+    player:addCurrency("scyld", scyldBounty)
```
Adding currency here and below?

cocosolos commented on scripts/globals/hunts.lua:
```diff
@@ -0,0 +1,1616 @@
+    local scyldBounty = zone[huntZone].hunt[huntId].bounty
+
+    player:messageSpecial(msg[5])
+    player:messageSpecial(msg[6], scyldBounty, scylds)
```
This message should be moved to after you distribute the currency, then use the updated total.

cocosolos commented on scripts/globals/hunts.lua:
```diff
@@ -1586,19 +1586,19 @@ if huntEntry then
-       player:addCurrency("scyld", bounty)
+       player:addCurrency("scyld", scyldBounty)
+       player:messageSpecial(msg[5])
+       player:messageSpecial(msg[6], scyldBounty, scylds)
```
This will still give the old total, go with this:
```lua
    -- scylds cap at 1000
    if player:getCurrency("scyld") + scyldBounty > 1000 then
       player:setCurrency("scyld", 1000)
    else
       player:addCurrency("scyld", scyldBounty)
    end
    player:messageSpecial(msg[5])
    player:messageSpecial(msg[6], scyldBounty, player:getCurrency("scyld"))
```

cocosolos commented on scripts/globals/hunts.lua:
```diff
@@ -0,0 +1,1615 @@
+
+-- accepting hunt
+if huntEntry then
+    player:messageSpecial(msg[1])
+    player:messageSpecial(msg[2], huntEntry.fee, scylds)
```
These should be moved to after the fee is applied, and updated with `player:getCurrency("scyld")`

Apocarypse committed:
Corrected issue where Hunt Registries were not spawning
The `status` column for Ru'lude, Tavnazia, and Windy [S] in npc_list.sql
had the wrong values and thus were not spawning. Changed from 2 to 0.

cocosolos reviewed (approved):
Tried out a few hunts, everything looks like it's working as it should. Good job!
